File: leetcode/algorithms/67_add_binary/main.dart
class AddBinary {
  /**
   * BigInt
   *
   * Complexities:
   *   N - Length of `a`
   *   M - Length of `b`
   *   - Time Complexity: O(max(N, M))
   *   - Space Complexity: O(max(N, M))
   */
  String addBinary(String a, String b) {
    return (BigInt.parse(a, radix: 2) + BigInt.parse(b, radix: 2)).toRadixString(2);
  }


  // Solution
  /**
   * Reverse Iteration
   *
   * Complexities:
   *   N - Length of `a`
   *   M - Length of `b`
   *   - Time Complexity: O(max(N, M))
   *   - Space Complexity: O(max(N, M))
   */
  String solution1(String a, String b) {
    StringBuffer sb = StringBuffer();

    int i = a.length - 1;
    int j = b.length - 1;
    int carry = 0;

    while (i >= 0 || j >= 0 || carry > 0) {
      int sum = carry;

      if (i >= 0) {
        if (a[i] == '1') sum++;
        i--;
      }

      if (j >= 0) {
        if (b[j] == '1') sum++;
        j--;
      }

      sb.write(sum % 2);

      carry = sum ~/ 2;
    }

    return sb.toString().split('').reversed.join('');
  }

  /**
   * BigInt
   *
   * Complexities:
   *   N - Length of `a`
   *   M - Length of `b`
   *   - Time Complexity: O(max(N, M))
   *   - Space Complexity: O(max(N, M))
   */
  String solution2(String a, String b) {
    BigInt numA = BigInt.parse(a, radix: 2);
    BigInt numB = BigInt.parse(b, radix: 2);

    return (numA + numB).toRadixString(2);
  }
}

File: leetcode/algorithms/67_add_binary/main.py
class AddBinary:
    """
    # BigInt
    #
    # Complexities:
    #   N - Length of `a`
    #   M - Length of `b`
    #   - Time Complexity: O(max(N, M))
    #   - Space Complexity: O(max(N, M))
    """
    def addBinary(self, a: str, b: str) -> str:
        return format(int(a, 2) + int(b, 2), "b")


    # Solution
    """
    # Solution 1
    #
    # Two Pointers & Carry
    #
    # Complexities:
    #   N - Length of `a`
    #   M - Length of `b`
    #   - Time Complexity: O(max(N, M))
    #   - Space Complexity: O(max(N, M))
    """
    def solution1(self, a: str, b: str) -> str:
        res = []
        carry = 0
        i, j = len(a) - 1, len(b) - 1

        while i >= 0 or j >= 0 or carry:
            total = carry

            if i >= 0:
                total += int(a[i])
                i -= 1

            if j >= 0:
                total += int(b[j])
                j -= 1

            res.append(str(total % 2))
            carry = total // 2

        return "".join(res[::-1])

    """
    # Solution 2
    #
    # Built-in function
    #
    # Complexities:
    #   N - Length of `a`
    #   M - Length of `b`
    #   - Time Complexity: O(max(N + M))
    #   - Space Complexity: O(max(N, M))
    """
    def solution2(self, a: str, b: str) -> str:
        return bin(int(a, 2) + int(b, 2))[2:]
